src/tests/integration/handle-cached.test.ts

```diff
@@ -0,0 +1,19 @@
+import { expect } from 'chai'
+import { createMocks } from './mocks'
+import { runBackground } from './steps/run-background'
+import { mount } from '../../popup/mount'
+
+describe('use existing handle if it is cached', () => {
```

Comment:
Ah, yeah this is a tricky thing about mocha, but makes sense once you understand it.

The `describe` blocks all run immediately when mocha is first invoked so that mocha "collects" all the it/before/after blocks it's supposed to run. So if you have something like 

```typescript
describe('foo', () => {
    before(() => console.log('This gets run later'))
    mount(mocks.chrome as any, mocks.popupWindow as any)
})
```

Then the code in the before gets run later. If you look at all the "steps" they are each declared inside `before` blocks. This is important because the mocks are also setup within a before block so if you don't have your code inside a before block it will run before the mocks have been set up, which is exactly what's happening with `mount(mocks.chrome as any, mocks.popupWindow as any)`

So the code below behaves closer to what you want

```typescript
import { expect } from 'chai'
import { createMocks } from './mocks'
import { runBackground } from './steps/run-background'
import { mountPopup } from './steps/mount-popup'

describe('use existing handle if it is cached', () => {
  const mocks = createMocks()

  runBackground(mocks, { alreadyAuthenticated: true })
  mountPopup(mocks, { alreadyAuthenticated: true })

  describe('foo', () => {
    it('should not cache the handle if it is already cached', () => {
      expect(mocks.chrome.storage.local.get).to.have.callCount(1)
      expect(mocks.chrome.storage.local.set).to.have.not.been.called
      // tslint:enable no-unused-expression
    })
  })
})
```

Reply:
1. What I don't understand is why would I want `runBackground` and `mountPop` to run when `mocha` collects the `describe`? I want them to run before I run my test. I want to make sure I have a "fresh" environment. When I try to move `runBackground` and `mountPop` inside a `before`, I get an error `ReferenceError: window is not defined`.

2. I don't think your suggestion will work anyway because `mountPop` sets the cache to be empty. I want to set the cache with a value before we mount the popup.

Reply:
I pushed some changes here which I think address these issues. `mountPopup` has its own describe block with its own before/after statements so we leverage those here.